constraints/recursiveCatEvals.js

```diff
@@ -53,29 +86,56 @@ var sCat = ["cp", "xp", "x0"];
  * ancestors before its maximality or minimality is used, your function will be
  * working with the correct values of isMin, isMax and parentCat.
  *
- * 7/29/19 refactor of an earlier version
+ * 10/26/20 update of an earlier version, now includes dummies.
  */
 
-function markMinMax(mytree){
+function markMinMax(mytree, options){
+	if(options.requireLexical){
+		mytree = createDummies(mytree, 'func');
+	}
+	if(options.requireOvertHead){
+		mytree = createDummies(mytree, 'silent');
+	}
+	return markMinMaxInner(mytree, options)
+}
+
+function markMinMaxInner(mytree, options){
 	/* If parentCat property is not already defined for this node, it is probably
 	 * the root node. Non-root nodes get the property parentCat when this node's
 	 * children are marked below.
 	 */
+	options = options || {};
+
 	if (!mytree.hasOwnProperty('parentCat')){
 		mytree.parentCat = "is root"; //marks the root node
 	}
 
-	//mark maximal nodes
-	mytree.isMax = (mytree.cat !== mytree.parentCat);
+	//Store the info of the most recent cat in order to skip over dummy nodes
+	//except when the dummy node's parent is ALSO a dummy node, then lastcat should be passed
+	//down dummy generation after dummy generation until a normal node is reached to inherit
+	//it as the parentCat.
+	if(mytree.cat === "dummy"){
+		mytree.isMax = false;
+		mytree.isMin = false;
+		if(mytree.parentCat !== "dummy"){
+			mytree.lastCat = mytree.parentCat;
+		}
+	} else {
+		//mark maximality and minimality for node
+		mytree.isMax = (mytree.cat !== mytree.parentCat);
+		mytree.isMin = isMinimal(mytree);
 
-	//mark minimality (relies on isMinimal above)
-	mytree.isMin = isMinimal(mytree);
+		//recall stored parentCat after dummies are skipped
+		if(mytree.parentCat === "dummy"){
+			mytree.parentCat = mytree.lastCat;
+		}
+	}
 
 	if(mytree.children && mytree.children.length){
 		for(var i = 0; i < mytree.children.length; i++){
-			var child = mytree.children[i];
-			child.parentCat = mytree.cat; // set the property parentCat
-			mytree.children[i] = markMinMax(mytree.children[i]);
+			mytree.children[i].parentCat = mytree.cat; // set the property parentCat
+			mytree.children[i].lastCat = mytree.lastCat; //pass on lastCat
```

Comment:
Given that you're recording lastCat on all the nodes, I'm not clear on why it's also needed as an argument to isMinimal. Can't isMinimal refer to node.lastCat? But perhaps I'm missing something.